src/Agent.Worker/JobExtension.cs

```diff
@@ -162,6 +164,30 @@ public async Task<List<IStep>> InitializeJob(IExecutionContext jobContext, Pipel
                         }
                     }
 
+                    //Check if we need to install old node runners
+                    var requiredOldNodeRunnersValue = AgentKnobs.InstallOldNodeRunners.GetValue(context).AsString();
+                    if (!string.IsNullOrEmpty(requiredOldNodeRunnersValue))
+                    {
+                        var requiredOldNodeRunners = requiredOldNodeRunnersValue.Split(';');
+                        foreach (var version in requiredOldNodeRunners)
+                        {
+                            var installer = new Pipelines.TaskStep()
+                            {
+                                Id = Guid.Parse("31C75B2B-BCDF-4706-8D7C-4DA6A1959BC5"),
+                                DisplayName = $"Install Node {version} runner",
+                                Reference = new Pipelines.TaskStepDefinitionReference()
+                                {
+                                    Id = Guid.Parse("31C75B2B-BCDF-4706-8D7C-4DA6A1959BC2"),
+                                    Name = "NodeTaskRunnerInstaller",
+                                    Version = "0.229.0",
+                                }
+                            };
+                            installer.Inputs.Add("runnerVersion", version);
```

Comment:
>installer [](http://example.com/codeflow?start=28&length=9)

this will slow down the pipelines for every run?  especially non-persistent agents.

Ideally (if we can) we'd handle this earlier; when the agent starts up, before it receives a job

Because 'Enable installation old Node Runners (node 6)' is the default, I think this is a blocker

Reply:
if the runner already installed, it will skip by saying - Node {version} runner already installed

Reply:
My concern is for non-persistent agents e.g. scale set/MMS, etc, all those would be impacted.  Can we install the node6 runner (if enabled) as early as possible e.g. when agent is registered?  We can also check when job starts in case the setting is changed.

In any case, we should evaluate the performance impact